File: c++/3laba_tests/src/simplyList.cpp
#include "simplyList.h"
#include <iostream>
using namespace std;

SimplyListNode::SimplyListNode(const string &value): data(value), next(nullptr){}

string SimplyListNode::getData() const {
    return data;
}

SimplyListNode *SimplyListNode::getNext() const {
    return next;
}

void SimplyListNode::setNext(SimplyListNode *node) {
    next = node;
}

SimplyList::SimplyList(): head(nullptr), tail(nullptr){}
SimplyList::~SimplyList() {
    while (head != nullptr) {
        const SimplyListNode* temp = head;
        head = head->getNext();
        delete temp;
    }
}

void SimplyList::addHead(const string &value) {
    if (value.empty()) {
        throw runtime_error("Значение пустое");
    }
    const auto newNode = new SimplyListNode(value);
    newNode->setNext(head);
    head = newNode;
    if (!tail) tail = newNode;
}

void SimplyList::addBack(const string &value) {
    if (value.empty()) {
        throw runtime_error("Значение пустое");
    }

    const auto newNode = new SimplyListNode(value);
    if (!tail) head = tail = newNode;
    else {
        tail->setNext(newNode);
        tail = newNode;
    }
}

void SimplyList::addAfterIndex(int index, const string &value) {
    if (index < 0) throw runtime_error("Индекс меньше 0");
    if (value.empty()) throw runtime_error("Значение пустое");

    if (head == nullptr) {
        if (index == 0) {
            head = new SimplyListNode(value);
            tail = head;
            return;
        } else {
            throw runtime_error("Список пуст, индекс недоступен");
        }
    }

    auto current = head;
    for (int i = 0; i < index && current != nullptr; i++) {
        current = current->getNext();
    }

    if (current == nullptr) {
        throw runtime_error("Индекс превышает размер списка");
    }

    const auto newNode = new SimplyListNode(value);
    newNode->setNext(current->getNext());
    current->setNext(newNode);

    if (newNode->getNext() == nullptr) {
        tail = newNode;
    }
}

void SimplyList::addBeforeIndex(int index, const string &value) {
    if (index < 0) throw runtime_error("Индекс меньше 0");
    if (value.empty()) throw runtime_error("Значение пустое");

    if (head == nullptr) {
        if (index == 0) {
            addHead(value);
            return;
        } else {
            throw runtime_error("Список пуст, индекс недоступен");
        }
    }

    if (index == 0) {
        addHead(value);
        return;
    }

    auto current = head;
    for (int i = 0; i < index - 1 && current != nullptr; i++) {
        current = current->getNext();
    }

    if (current == nullptr) {
        throw runtime_error("Индекс превышает размер списка");
    }

    const auto newNode = new SimplyListNode(value);
    newNode->setNext(current->getNext());
    current->setNext(newNode);

    if (newNode->getNext() == nullptr) {
        tail = newNode;
    }
}

void SimplyList::dellHead() {
    if (head == nullptr) throw runtime_error("Список пустой");

    const auto temp = head;
    head = head->getNext();

    if (temp == tail) {
        tail = nullptr;
    }

    delete temp;
}

void SimplyList::dellBack() {
    if (head == nullptr) throw runtime_error("Список пустой");

    if (head == tail) {
        delete head;
        head = tail = nullptr;
        return;
    }

    auto current = head;
    while (current ->getNext() != tail) current = current ->getNext();

    delete tail;
    tail = current;
    tail -> setNext(nullptr);
}

void SimplyList::dellAfterIndex(int index) {
    if (head == nullptr) throw runtime_error("Список пустой");
    if (index < 0) throw runtime_error("Индекс меньше 0");

    auto current = head;
    for (int i = 0; i < index && current != nullptr; i++) {
        current = current ->getNext();
    }

    if (current == nullptr || current->getNext() == nullptr) throw runtime_error("Индекс превышает размер списка");

    const auto temp = current->getNext();
    current->setNext(temp->getNext());

    if (temp == tail) {
        tail = current;
    }
    delete temp;
}

void SimplyList::dellBeforeIndex(int index) {
    if (head == nullptr) throw runtime_error("Список пустой");
    if (index < 0) throw runtime_error("Индекс меньше 0");
    if (index == 0) throw runtime_error("Нельзя удалить элемент перед индексом 0");

    if (index == 1) {
        dellHead();
        return;
    }


    auto current = head;
    for (int i = 0; i < index - 2 && current != nullptr; i++) {
        current = current ->getNext();
    }

    if (current == nullptr || current->getNext() == nullptr) throw runtime_error("Индекс превышает размер списка");

    const auto temp = current->getNext();
    current->setNext(temp->getNext());

    if (temp == tail) {
        tail = current;
    }
    delete temp;
}

void SimplyList::dellValue(const string &value) {
    if (head == nullptr) throw runtime_error("Список пустой");
    if (value.empty()) throw runtime_error("Значение пустое");

    if (head->getData() == value) {
        const auto temp = head;
        head = head->getNext();
        if (head == nullptr) tail = nullptr;
        delete temp;
        return;
    }

    auto* current = head;
    while (current->getNext() != nullptr && current->getNext()->getData() != value) {
        current = current->getNext();
    }

    if (current ->getNext() == nullptr) throw runtime_error(value + " отсутствует в списке");

    const auto temp = current->getNext();
    current->setNext(temp->getNext());

    if (temp == tail) {
        tail = current;
    }

    delete temp;
}

bool SimplyList::searchValue(const string &value) const {
    if (head == nullptr) throw runtime_error("Список пустой");
    if (value.empty()) throw runtime_error("Значение пустое");

    auto current = head;
    while (current != nullptr) {
        if (current->getData() == value) return true;
        current = current->getNext();
    }
    return false;
}

void SimplyList::print() const {
    if (head == nullptr) {
        cout << "(Список пустой)" << endl;
        return;
    }
    auto current = head;
    while (current != nullptr) {
        cout << current->getData();
        if (current->getNext() != nullptr) {
            cout << " -> ";
        }
        current = current->getNext();
    }
    cout << endl;
}

SimplyListNode *SimplyList::getHead() const {
    return head;
}

File: rust/src/simply_list.rs
use std::fmt;
use serde::{Deserialize, Serialize};

#[derive(Clone, Serialize, Deserialize)]
pub struct SimplyListNode {
    data: String,
    next: Option<Box<SimplyListNode>>,
}

impl SimplyListNode {
    pub fn new(value: &str) -> Self {
        SimplyListNode {
            data: value.to_string(),
            next: None,
        }
    }

    pub fn get_data(&self) -> &str {
        &self.data
    }

    pub fn get_next(&self) -> Option<&SimplyListNode> {
        self.next.as_deref()
    }

    pub fn set_next(&mut self, node: Option<Box<SimplyListNode>>) {
        self.next = node;
    }
}
#[derive(Clone, Serialize, Deserialize)]
pub struct SimplyList {
    pub(crate) head: Option<Box<SimplyListNode>>,
}

impl SimplyList {
    pub fn new() -> Self {
        SimplyList { head: None }
    }

    pub fn get_head(&self) -> Option<&SimplyListNode> {
        self.head.as_deref()
    }

    pub fn add_head(&mut self, value: &str) -> Result<(), &'static str> {
        if value.is_empty() {
            return Err("Значение пустое");
        }

        let mut new_node = Box::new(SimplyListNode::new(value));
        new_node.set_next(self.head.take());
        self.head = Some(new_node);
        Ok(())
    }

    pub fn add_back(&mut self, value: &str) -> Result<(), &'static str> {
        if value.is_empty() {
            return Err("Значение пустое");
        }

        if self.head.is_none() {
            return self.add_head(value);
        }

        let mut current = &mut self.head;
        while current.as_ref().unwrap().next.is_some() {
            current = &mut current.as_mut().unwrap().next;
        }

        current.as_mut().unwrap().set_next(Some(Box::new(SimplyListNode::new(value))));
        Ok(())
    }

    pub fn add_after_index(&mut self, index: usize, value: &str) -> Result<(), &'static str> {
        if value.is_empty() {
            return Err("Значение пустое");
        }

        if self.head.is_none() {
            if index == 0 {
                return self.add_head(value);
            } else {
                return Err("Список пуст, индекс недоступен");
            }
        }

        let mut current = &mut self.head;
        for _ in 0..index {
            match current {
                Some(node) => {
                    current = &mut node.next;
                }
                None => return Err("Индекс превышает размер списка"),
            }
        }

        if let Some(node) = current {
            let mut new_node = Box::new(SimplyListNode::new(value));
            new_node.set_next(node.next.take());
            node.set_next(Some(new_node));
            Ok(())
        } else {
            Err("Индекс превышает размер списка")
        }
    }

    pub fn add_before_index(&mut self, index: usize, value: &str) -> Result<(), &'static str> {
        if value.is_empty() {
            return Err("Значение пустое");
        }

        if index == 0 {
            return self.add_head(value);
        }

        self.add_after_index(index - 1, value)
    }

    pub fn dell_head(&mut self) -> Result<(), &'static str> {
        match self.head.take() {
            Some(node) => {
                self.head = node.next;
                Ok(())
            }
            None => Err("Список пустой"),
        }
    }

    pub fn dell_back(&mut self) -> Result<(), &'static str> {
        if self.head.is_none() {
            return Err("Список пустой");
        }

        if self.head.as_ref().unwrap().next.is_none() {
            return self.dell_head();
        }

        let mut current = &mut self.head;
        while current.as_ref().unwrap().next.as_ref().unwrap().next.is_some() {
            current = &mut current.as_mut().unwrap().next;
        }

        current.as_mut().unwrap().set_next(None);
        Ok(())
    }

    pub fn dell_after_index(&mut self, index: usize) -> Result<(), &'static str> {
        if self.head.is_none() {
            return Err("Список пустой");
        }

        let mut current = &mut self.head;
        for _ in 0..index {
            match current {
                Some(node) => {
                    current = &mut node.next;
                }
                None => return Err("Индекс превышает размер списка"),
            }
        }

        if let Some(node) = current {
            if node.next.is_some() {
                let next_node = node.next.take().unwrap();
                node.set_next(next_node.next);
                Ok(())
            } else {
                Err("Индекс превышает размер списка")
            }
        } else {
            Err("Индекс превышает размер списка")
        }
    }

    pub fn dell_before_index(&mut self, index: usize) -> Result<(), &'static str> {
        if index == 0 {
            return Err("Нельзя удалить элемент перед индексом 0");
        }

        if index == 1 {
            return self.dell_head();
        }

        self.dell_after_index(index - 2)
    }

    pub fn dell_value(&mut self, value: &str) -> Result<(), &'static str> {
        if value.is_empty() {
            return Err("Значение пустое");
        }

        if self.head.is_none() {
            return Err("Список пустой");
        }

        if self.head.as_ref().unwrap().data == value {
            return self.dell_head();
        }

        let mut current = &mut self.head;
        while current.as_ref().unwrap().next.is_some() {
            if current.as_ref().unwrap().next.as_ref().unwrap().data == value {
                break;
            }
            current = &mut current.as_mut().unwrap().next;
        }

        if let Some(node) = current {
            if node.next.is_some() {
                let next_node = node.next.take().unwrap();
                node.set_next(next_node.next);
                Ok(())
            } else {
                Err("Значение отсутствует в списке")
            }
        } else {
            Err("Значение отсутствует в списке")
        }
    }

    pub fn search_value(&self, value: &str) -> Result<bool, &'static str> {
        if value.is_empty() {
            return Err("Значение пустое");
        }

        if self.head.is_none() {
            return Err("Список пустой");
        }

        let mut current = &self.head;
        while let Some(node) = current {
            if node.data == value {
                return Ok(true);
            }
            current = &node.next;
        }
        Ok(false)
    }

    pub fn print(&self) {
        if self.head.is_none() {
            println!("(Список пустой)");
            return;
        }

        let mut current = &self.head;
        while let Some(node) = current {
            print!("{}", node.data);
            if node.next.is_some() {
                print!(" -> ");
            }
            current = &node.next;
        }
        println!();
    }
}

impl Drop for SimplyList {
    fn drop(&mut self) {
        while self.dell_head().is_ok() {}
    }
}

impl fmt::Display for SimplyList {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        if self.head.is_none() {
            return write!(f, "(Список пустой)");
        }

        let mut current = &self.head;
        let mut first = true;

        while let Some(node) = current {
            if !first {
                write!(f, " -> ")?;
            }
            write!(f, "{}", node.data)?;
            current = &node.next;
            first = false;
        }

        Ok(())
    }
}